feat: integrate unstructured module

Wendong-Fan opened the pull request:
## Description

integrate the unstructured module to CAMEL

## Motivation and Context

Data ETL(extract, transform, and load)
Integrate [[unstructured](https://unstructured.io/)](https://[unstructured](https://unstructured.io/).io/) with CAMEL

[unstructured](https://unstructured.io/) supports.txt, .docx, .pptx, .jpg, .png, .eml, .html, and .pdf, integrated with chunking and embedding

 `close #365 ` 

- [X] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation (update in the documentation)
- [X] Example (update in the folder of example)

## Implemented Tasks

- [x] Unstructured version check
- [x] File parse
- [x] Data clean
- [x] Data extract
- [x] Data stage
- [x] Data chunck
- [ ] Embedding

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

FUYICC commented:
When I run examples\file_io\unstructured_modules_example.py, selecting 1 prompts me that “can't find the file”. I changed the path and it worked. I'm not sure if there is a better path or we have no exact path for Github, what are your thoughts? @Wendong-Fan

FUYICC commented on test/file_io/test_unstructured_io.py:
```diff
@@ -0,0 +1,126 @@
+    # Test with an invalid extract option (should raise ValueError)
+    options = {"invalid_extracting_option": {}}
+    with pytest.raises(ValueError):
+        unstructured_instance.extract_data_from_text("Test Text", options)
```
Please note that the inconsistency in the order of input between the text and operation type mentioned earlier has led to misunderstandings. For both the "extract_data_from_text" and the next "stage_elements" functions, the expected order of inputs is to first provide the operation type as a string, followed by the text. Hence, the reason why "Text Test" passes in pytest is because it is not present in the existing extraction_functions.

If I didn't misunderstanding these funcionalities, I suggest modifying this as follows:
```
option = "invalid_extracting_option"
with pytest.raises(ValueError):
    unstructured_instance.extract_data_from_text(option, "Test `Text")
```

Wendong-Fan replied:
you are right, here should be a string input, have fixed

FUYICC reviewed (changes requested):
I just tried it briefly and noticed these issues. I haven't thoroughly checked the functionality yet.

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,351 @@
+        r"""Loads a file or a URL and parses its contents as unstructured data.
+
+        Args:
+            input_path (str): Path to the file or URL to be parsed.
```
added notes in doc

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,351 @@
+                raise Exception(
+                    "Failed to parse the unstructured file.") from e
+
+    def clean_text_data(self, text: str, options: Dict[str, Any]) -> str:
```
changed order of "clean_text_data" function input

Wendong-Fan commented on test/file_io/test_unstructured_io.py:
```diff
@@ -0,0 +1,126 @@
+    assert extracted_email == ["example@email.com"]
+
+    # Test with an invalid extract option (should raise ValueError)
+    options = {"invalid_extracting_option": {}}
```
for "clean_text_data" function, users may want to clean different kinds of elements at one time, by using Dict can let users do it with one function running. For other functions like extract and stage, I think users may not do multi-commands at one time.

Wendong-Fan commented:
> I just tried it briefly and noticed these issues. I haven't thoroughly checked the functionality yet.

Hi @FUYICC , thank you so much for your valuable review, I have made some changes based on your feedback, feel free to check it, thanks!

FUYICC reviewed (approved):
Thanks a lot @Wendong-Fan! I finished a new review for your rectified work, it looks good right now.

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,351 @@
+            from unstructured.__version__ import __version__
+
+        except ImportError as e:
+            raise ImportError("unstructured package is not installed.") from e
```
solved

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,351 @@
+        installed_ver = version.parse(__version__)
+
+        if installed_ver < min_ver:
+            raise ValueError(f"unstructured>={min_version} required, "
```
solved

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,351 @@
+        return extraction_functions[extract_type](text, **kwargs)
+
+    def stage_elements(self, stage_type: str, elements: List[Any],
+                       **kwargs) -> Union[str, List[Dict], pd.DataFrame]:
```
solved

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,351 @@
+        Args:
+            stage_type (str): Type of staging to perform. Supported types:
+                            'csv', 'dataframe', 'dict', 'prodigy',
+                            'baseplate', 'datasaur', 'label_box', 'weaviate'.
```
solved

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,360 @@
+            By default we use the basic "unstructured" library,
+            If you are processing document types beyond the basics,
+            you can install the necessary extras like:
+            pip install "unstructured[docx,pptx]"
```
solved

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,360 @@
+        r"""Chunks elements by titles.
+
+        Args:
+            chunk_type (str): Type chunk going to apply.
```
solved

Wendong-Fan commented on pyproject.toml:
```diff
@@ -24,7 +24,7 @@ repository = "https://github.com/camel-ai/camel"
 
 [tool.poetry.dependencies]
-python = "^3.8.1"
+python = ">=3.8.1,<3.12"
```
one of unstructured io's dependencies requires this Python library set, or it will have an error

dandansamax replied:
btw, Qdrant backend requires `python<3.12`, too.

lightaime commented:
How should we combine this and https://github.com/camel-ai/camel/blob/master/camel/functions/data_io_functions.py? Should we move it to the `functions` (or `tools`) that we may need to rename?

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,360 @@
+            the names of the cleaning functions, and the
+            values should be dictionaries containing the
+            parameters for each function.
+            text (str): The text to be cleaned.
```
yes we need to use tuple to guarantee the order.

I defined a `default_clean_options`

```
            clean_options = [
                ("replace_unicode_quotes", {}),
                ("clean_non_ascii_chars", {}),
                ("group_broken_paragraphs", {}),
                ("clean_extra_whitespace", {}),
            ]
```

I think these 4 clean methods can be applied in most scenarios, but other clean methods should be taken case by case, let me know if you have any suggestions

lightaime commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,411 @@
+            raise ValueError(f"Require `unstructured>={min_version}`, "
+                             f"you have {__version__}.")
+
+    def parse_file_or_url(self, input_path: str) -> Any:
```
Should me have an argument `return_str` and return `str` by default? That may be easy for the agent to use it a tool. Also it would be easier for the users who are not familiar with `unstructured`.

Wendong-Fan replied:
great point, have added the argument

lightaime reviewed (commented):
I still think we should be `unstructured_io.py` to `functions/data_tools/`. WDYT?

Wendong-Fan commented on camel/file_io/unstructured_io.py:
```diff
@@ -0,0 +1,411 @@
+    def clean_text_data(
+        self, text: str,
+        clean_options: Optional[List[Tuple[str, Dict[str,
+                                                     Any]]]] = None) -> str:
```
solved

Wendong-Fan commented:
> How should we combine this and https://github.com/camel-ai/camel/blob/master/camel/functions/data_io_functions.py? Should we move it to the `functions` (or `tools`) that we may need to rename? @dandansamax any opinion on this?

@lightaime I have moved unstructured module into functions and renamed `data_io_functions.py` into `base_io_functions.py`, let me know if you have other idea~

zechengz commented on camel/functions/unstructured_io_fuctions.py:
```diff
@@ -0,0 +1,422 @@
+
+
+class UnstructuredModules:
+    UNSTRUCTURED_MIN_VERSION = "0.10.30"  # Define the minimum version
```
nit: just use `version.Version('0.10.30')` instead?

zechengz commented on camel/functions/unstructured_io_fuctions.py:
```diff
@@ -0,0 +1,422 @@
+            raise ValueError(f"Require `unstructured>={min_version}`, "
+                             f"you have {__version__}.")
+
+    def parse_file_or_url(
```
Any reason why we make file and url use the same parse function? Maybe it's better to separate parse file and url?

Wendong-Fan replied:
using a single function to handle both files and url simplifies the interface for the user, they will not need to worry about choosing different methods for files or url. I think it's also a better approach to get a more unified interface

zechengz commented on camel/functions/unstructured_io_fuctions.py:
```diff
@@ -0,0 +1,422 @@
+    def clean_text_data(
+        self,
+        text: str,
+        clean_options: Optional[List[Tuple[str, Dict[str, Any]]]] = None,
```
Maybe it's better to just let user pass a list of `Callable` instead of we fix the choices of options in the function. Similar suggestion for following functions. WDYT?

lightaime replied:
I think `Callable` is a good idea. But that will require users to learn `unstructured io` and hard to LLM agents to call this function. I think fix options would be sufficient for now.

zechengz commented on camel/functions/unstructured_io_fuctions.py:
```diff
@@ -0,0 +1,422 @@
+        from urllib.parse import urlparse
+
+        # Check installed unstructured version
+        self.ensure_unstructured_version(self.UNSTRUCTURED_MIN_VERSION)
```
I think we can just validate or verify the version once during `__init__`

lightaime commented:
> added remote data source connectors including s3, azure, github, slack, discord @lightaime

Do you want to add some tests on these connectors or later?
